File: js/easy/num-palindrome.js
// Given an integer x, return true if x is palindrome integer.

// An integer is a palindrome when it reads the same backward as forward.

// For example, 121 is a palindrome while 123 is not.

const isPalindrome = (num) => {
  // All negative nos are not palindrom, all nos ending with 0 are not palindrome
  // All single digit nos including 0 is palindrome
  if (num < 0 || (num % 10 == 0 && num != 0)) return false;
  if (num == 0 || num < 10) return true;
  let numStr = String(num);
  let strLen = numStr.length;
  for (let i = 0; i < strLen / 2; i++) {
    if (numStr[i] !== numStr[strLen - (i + 1)]) {
      return false;
    }
    return true;
  }
};

const isPalindromeWithoutConvertingToString = (num) => {
  // All negative nos are not palindrom, all nos ending with 0 are not palindrome
  // All single digit nos including 0 is palindrome
  if (num < 0 || (num % 10 == 0 && num != 0)) return false;
  if (num == 0 || num < 10) return true;
  // We split the number in half and then we
  let revertedNumber = 0;
  while (num > revertedNumber) {
    revertedNumber = Math.trunc(revertedNumber * 10 + (num % 10));
    num = Math.trunc(num / 10);
    console.log(revertedNumber, num, "here");
  }
  // When the length is an odd number, we can get rid of the middle digit by revertedNumber/10
  // For example when the input is 12321, at the end of the while loop we get x = 12, revertedNumber = 123,
  // since the middle digit doesn't matter in palidrome(it will always equal to itself), we can simply get rid of it.
  console.log(revertedNumber);
  console.log(num);
  return num == revertedNumber || num == Math.trunc(revertedNumber / 10);
};

console.log(isPalindromeWithoutConvertingToString(98989));
// console.log(isPalindromeWithoutConvertingToString(123));

File: py/easy/num-palindrome.py
'''
// Given an integer x, return true if x is palindrome integer.

// An integer is a palindrome when it reads the same backward as forward.

// For example, 121 is a palindrome while 123 is not.
'''

def isPalindromeWithoutString(x: int) -> bool:
    if x < 0 or (x %10 ==0 and x !=0):
        return False
    if x == 0 or x < 10:
        return True
    reverted_num = 0
    while(reverted_num < x):
        reverted_num = (reverted_num * 10) + (x % 10)
        x = int(x / 10)

    return x == reverted_num or x == int(reverted_num / 10)


def isPalindrome(x: int) -> bool:
    if x < 0 or (x %10 ==0 and x !=0):
        return False
    if x == 0 or x < 10:
        return True
    num_str = str(x)
    num_len = len(num_str)
    for i in range(int(num_len/2)):
        if num_str[i] != num_str[-(i+1)]:
            return False
    return True

print(isPalindromeWithoutString(45954))